### PR TITLE
feat(apps/legacy-api): perf improvement with caching

### PR DESCRIPTION
**Pull request by eli-lim**
What this PR does / why we need it:
- Introduces block-level + transaction-level swap caching for getsubgraphswaps query path
- Recent blocks (within chain height - 2) are not cached as they could be invalidated. However, transaction-level swap cache enables us to call AccountHistory rpc at most once for any given txn, hence largely side-stepping the performance bottleneck.

#### 1. Block-level caching of pool swaps needed to get response time in the milliseconds range
- Unfortunately, it's not enough to just cache by txid of the pool swap. Although the bottleneck of accounthistory rpc is side-stepped with this, the times from traversing blocks and their transactions add up.
- Hence, I tried caching the swaps by block hash instead, although we probably shouldn't cache recent blocks since they could be invalidated / more swaps could be added to them. 
- This was great for performance!

#### 2. Fixing pagination and making it work with this caching approach
- Pagination is currently wrongly implemented. Every returned `next` token when used to hit the endpoint again will **skip the block** even if there are pool swaps remaining in the block.
  ```
  blockchain = [ 
    block 2 [{ swap D }, ..., { swap C }, ..., { swap B }]
    block 1 [{ swap A }, ...]
  ]

  /getsubgraphswaps?limit=2 
    - gives: { swaps: [{ swap D }, { swap C }], page: XXX }

  /getsubgraphswaps?limit=2&next=XXX
    - should give:  [{ swap B }, { swap A }]
    - but gives:  [{ swap A }]
  ```
- This is due to the `next.height` value being returned to the client as the current block's height, say 2. So in the next pagination call, the provided `next.height` of 2 is passed to the node rpc call, thereby skipping block 2 (this is expected behaviour).
- We therefore should +1 to `next.height` if the swap **isn't the last swap in the block's swaps**.

Which issue(s) does this PR fixes?:
- Part of https://github.com/DeFiCh/jellyfish/issues/1024

# Sanity checks
- [x] Responses are the same as https://api.defichain.com/v1/getsubgraphswaps
- [x] Pagination is correct

### CODE REVIEW

**Review by fuxingloh — COMMENTED** [review]
looks simple 🚀

How's the performance?

**Comment by eli-lim** [comment]
> looks simple 🚀
> 
> How's the performance?

Pretty good - about 200ms for 100 swaps (once cache is warmed up)

File: apps/legacy-api/src/controllers/PoolPairController.ts [hunk 1]
```diff
@@ -513,3 +647,13 @@
 
   return undefined
 }
+
+export function encodeBase64 (next: NextToken): string {
+  return Buffer.from(JSON.stringify(next), 'utf8').toString('base64url')
```
Comment (weiyuan95) [inline apps/legacy-api/src/controllers/PoolPairController.ts:652]: `toString(base64url) was introduced in v14.18` ref: https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V14.md#14.18.0 so we'll need to set a dependency on this version in `jellyfish/package.json` (don't ask me how I realised this)

**Review by weiyuan95 — CHANGES_REQUESTED** [review]

File: apps/legacy-api/src/controllers/PoolPairController.ts [hunk 2]
```diff
@@ -287,6 +373,54 @@
   }
 }
 
+/**
+ * Hits PoolPairController with requests to enable pre-fetching behaviour,
+ * improving performance on queries
+ */
+@Injectable()
+export class SwapCacheFiller {
```
[the remark this answers is not in the document]
Comment (eli-lim) [reply apps/legacy-api/src/controllers/PoolPairController.ts:381]: Same area of concern, hence keeping it where it is

File: apps/legacy-api/__tests__/controllers/PoolPairController.test.ts [hunk 3]
```diff
@@ -262,12 +295,192 @@ describe('getsubgraphswaps', () => {
...
+
+    expect(swap1.data.swaps[0]).toStrictEqual(swap1And2.data.swaps[0])
+    expect(swap2.data.swaps[0]).toStrictEqual(swap1And2.data.swaps[1])
+  })
+})
+
+describe('getsubgraphswaps - relying on caching', () => {
+  beforeAll(async () => {
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    async function relyOnProduction (): Promise<void> {
```
[the remark this answers is not in the document]
Comment (eli-lim) [reply apps/legacy-api/__tests__/controllers/PoolPairController.test.ts:333]: No - but it's useful for local sanity testing to actually see the production data being served.

File: apps/legacy-api/src/controllers/PoolPairController.ts (hunk 1 above)
Comment (eli-lim) [reply apps/legacy-api/src/controllers/PoolPairController.ts:652]: Lemme guess - you were running with < v14.18 locally? LOL
So you're suggesting we bump https://github.com/DeFiCh/jellyfish/blob/ee2016fd86e172a53fd5ef5a2393173ce8fd9807/package.json#L17 to `>= 14.18` ?

If you have `nvm` installed, IDEA ide should automagically pick the node runtime when running anything in the project

**Review by weiyuan95 — APPROVED** [review]
:shipit:

File: apps/legacy-api/src/controllers/PoolPairController.ts (hunk 1 above)
Comment (weiyuan95) [reply apps/legacy-api/src/controllers/PoolPairController.ts:652]: We don't have a .npmrc which enforces the version of node, so marking this as resolved

**Review by fuxingloh — APPROVED** [review]

**Pull request merged**